jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleHelper.java

```diff
@@ -0,0 +1,699 @@
+			if (null == bsm) {
+				/*[MSG "K05cd", "unable to resolve 'bootstrap_method_ref' in '{0}' at index {1}"]*/
+				throw new NullPointerException(Msg.getString("K05cd", classObject.toString(), bsmIndex)); //$NON-NLS-1$
+			}
+			Object[] staticArgs = new Object[BSM_OPTIONAL_ARGUMENTS_START_INDEX + bsmArgCount];
+			/* Mandatory arguments */
+			staticArgs[BSM_LOOKUP_ARGUMENT_INDEX] = lookup;
+			staticArgs[BSM_NAME_ARGUMENT_INDEX] = name;
+			staticArgs[BSM_TYPE_ARGUMENT_INDEX] = type;
+		
+			/* Static optional arguments */
+			int bsmTypeArgCount = bsm.type().parameterCount();
+			for (int i = 0; i < bsmArgCount; i++) {
+				staticArgs[BSM_OPTIONAL_ARGUMENTS_START_INDEX + i] = getAdditionalBsmArg(access, internalRamClass, classObject, bsm, bsmArgs, bsmTypeArgCount, i);
+			}
+
+/*[IF Java11]*/
+		/* JVMS JDK11 5.4.3.6 Dynamically-Computed Constant and Call Site Resolution
+		 * requires that exceptions from BSM invocation be wrapped in a BootstrapMethodError
+		 * unless the exception thrown is a sub-class of Error.
+		 * Exceptions thrown before invocation should be passed through unwrapped.
+		 */
+		try {
+/*[ENDIF]*/
+			CallSite cs = (CallSite)invokeBsm(bsm, staticArgs);
+			if (cs != null) {
+				MethodType callsiteType = cs.type();
+				if (callsiteType != type) {
+					throw WrongMethodTypeException.newWrongMethodTypeException(type, callsiteType);
+				}
+				result = cs.dynamicInvoker();
+			} 
+			/*[IF Java11]*/
+			else {
+				/* The result of the resolution of a dynamically-computed call site must not be null. */
+				/*[MSG "K0A02", "Bootstrap method returned null."]*/
+				throw new ClassCastException(Msg.getString("K0A02")); //$NON-NLS-1$
+			}
+			/*[ENDIF]*/
+		} catch(Throwable e) {
+
+			/*[IF Sidecar19-SE]*/
+			if (e instanceof Error) {
+				throw e;
+			}
+			/*[ENDIF]*/
+
+			if (type == null) {
+				throw new BootstrapMethodError(e);
+			}
+			
+			/* create an exceptionHandle with appropriate drop adapter and install that */
+			try {
+				MethodHandle thrower = MethodHandles.throwException(type.returnType(), BootstrapMethodError.class);
+				MethodHandle constructor = MethodHandles.Lookup.IMPL_LOOKUP.findConstructor(BootstrapMethodError.class, MethodType.methodType(void.class, Throwable.class));
+				result = MethodHandles.foldArguments(thrower, constructor.bindTo(e));
+				result = MethodHandles.dropArguments(result, 0, type.parameterList()); 
+			} catch (IllegalAccessException iae) {
+				throw new Error(iae);
+			} catch (NoSuchMethodException nsme) {
+				throw new Error(nsme);
+			}
+		}
+		
+		return result;
+/*[ENDIF] OPENJDK_METHODHANDLES*/
+	}
+
+	@SuppressWarnings("unused")
+	@VMCONSTANTPOOL_METHOD
+	private static final MethodHandle sendResolveMethodHandle(
+			int cpRefKind,
+			Class<?> currentClass,
+			Class<?> referenceClazz,
+			String name,
+			String typeDescriptor,
+			ClassLoader loader) throws Throwable {
+/*[IF OPENJDK_METHODHANDLES]*/
```

Comment:
Why do we have both `sendResolveMethodHandle` code for OpenJDK MHs and a new `resolveMethodHandle` method?  Is there a relationship between them?  Should there be?

Reply:
`resolveMethodHandle` generates the bytecodes for a MethodHandle (MH) to be invoked.
`sendResolveMethodHandle` stays the same, and it finds the MH based on the MH CP Ref kind.
+@fengxue-IS, to validate the above statements.

Reply:
`resolveMethodHandle` is used to generate memberName adapter for a `invokehandle` bytecode (invocation),
`sendResolveMethodHandle` is used for ldc bytecode that loads a MH (constant loading)

Reply:
Can one of them be renamed to make the different use cases clear?  It would also be good to comment them with this information so it's evident from the code when / why each should be used

Reply:
`resolveMethodHandle` renamed to `linkCallerMethod`.